### ISSUE TITLE
How do I catch OOM exceptions

### ISSUE THREAD

**Issue by coder-tq** [caoccao/Javet#258]
When I execute such a script

Let cc=[... new Array (1000000000). keys()];

The v8 engine seems to have experienced an OOM, but Javet did not throw a JavetOOM exception, but directly caused the Java process to crash. How do I capture the OOM information of the v8 engine and keep the Java process running

**Comment by caoccao** [comment]
Thank you for reporting this issue. It is not recommended to do that in Javet because there is no **sparse array** in Java so that Javet directly maps JS array to Java `ArrayList`.

Regarding the V8 running out of memory, I think capturing that event is not implemented yet. I'll fix it later.

**Comment by caoccao** [comment]
By the way, I tried that in Node.js. Node.js crashes as well. It doesn't even have enough time printing the whole call stack before being killed.

![image](https://github.com/caoccao/Javet/assets/17514279/7a0cf8bf-7a84-4e78-b723-5d3aa14fd49e)

**Comment by coder-tq** [comment]
> Thank you for reporting this issue. It is not recommended to do that in Javet because there is no **sparse array** in Java so that Javet directly maps JS array to Java `ArrayList`.
> 
> Regarding the V8 running out of memory, I think capturing that event is not implemented yet. I'll fix it later.

I will try to avoid using such syntax, but since our platform is user-facing, we cannot perfectly control user input, so we need to address this issue.

**Comment by caoccao** [comment]
> I will try to avoid using such syntax, but since our platform is user-facing, we cannot perfectly control user input, so we need to address this issue.

Agree. Anything that results in coredump needs to be addressed and fixed. Please be aware that it might take a long time as the nature of the tech challenge.

**Comment by caoccao** [comment]
I spent a few hours analyzing the source code of Node.js and V8. Unfortunately, `OS.Abort()` is hardcoded in V8 after OOM. If you try the same in Chrome console, that Chrome process crashes as well.

There's no way of turning it off unless there is a patch to V8 at source code level. Patching V8 is not one of the goals in Javet. If you really need to take complete control after OOM, a private build will be the solution. Please let me know if you have any questions.

**Comment by coder-tq** [comment]
I collect all the created V8 engines, and check them with a timer task every 100ms. If their memory usage exceeds the threshold I've set, I use the 'terminateExecution' method to stop execution. This successfully prevents process crashes.

**Comment by caoccao** [comment]
> I collect all the created V8 engines, and check them with a timer task every 100ms. If their memory usage exceeds the threshold I've set, I use the 'terminateExecution' method to stop execution. This successfully prevents process crashes.

That sounds like a good feature because `JavetEngineGuard` is the one babysitting the `JavetEngine`. I think that could justify as a feature request. What do you think?

**Comment by coder-tq** [comment]
> > I collect all the created V8 engines, and check them with a timer task every 100ms. If their memory usage exceeds the threshold I've set, I use the 'terminateExecution' method to stop execution. This successfully prevents process crashes.
> 
> That sounds like a good feature because `JavetEngineGuard` is the one babysitting the `JavetEngine`. I think that could justify as a feature request. What do you think?

I think this is a good idea. Currently, JavetEngineGuard can only limit the execution time of the engine. More resource limitations could be added, such as memory, CPU usage, etc.

**Comment by caoccao** [comment]
> I think this is a good idea. Currently, JavetEngineGuard can only limit the execution time of the engine. More resource limitations could be added, such as memory, CPU usage, etc.

Sounds good.

**Comment by coder-tq** [comment]
> I collect all the created V8 engines, and check them with a timer task every 100ms. If their memory usage exceeds the threshold I've set, I use the 'terminateExecution' method to stop execution. This successfully prevents process crashes.

Unfortunately, when I fetch the memory information of V8Runtime at a high frequency, if the V8Runtime is already closed, a coredump will occur. However, during the monitoring process, there is always a time difference between checking whether the engine is closed and fetching memory information. I am using version 1.1.0 of Javet. I wonder if higher versions have fixed this issue?

**Comment by caoccao** [comment]
Yesterday I published [Break Serverless in Node.js Easily](https://sjtucaocao.medium.com/break-serverless-in-node-js-easily-c73b8c9ca13) in which I mentioned:

> The higher frequency it profiles with, the more CPU cycles does it require, the less crash will happen.

This is not a permanent solution.

I think the right solution would be proactively validating the array length while expanding the array. That would require a code change to V8 and I guess that would come with a considerable performance overhead because the validation could happen per `push()`.

**Comment by caoccao** [comment]
> I wonder if higher versions have fixed this issue?

No, that hasn't been changed.

**Comment by caoccao** [comment]
This issue has finally been fixed in v4.1.6. Please refer to this [commit](https://github.com/caoccao/Javet/commit/3aa5525a1d168d2940dd02a86180251ca87e9275) for detail.
